packages/alfa-dom/src/tables/structure.ts

```diff
@@ -0,0 +1,353 @@
+import {Mapper} from "@siteimprove/alfa-mapper";
+import {clamp} from "@siteimprove/alfa-math";
+import {None, Option, Some} from "@siteimprove/alfa-option";
+import {Parser} from "@siteimprove/alfa-parser";
+import {Predicate} from "@siteimprove/alfa-predicate";
+import {Iterable} from "@siteimprove/alfa-iterable";
+import {Err, Ok, Result} from "@siteimprove/alfa-result";
+import {test} from "@siteimprove/alfa-test";
+import {Attribute, Element, Namespace, Node} from "..";
+import isElement = Element.isElement;
+
+const { and, equals, property } = Predicate;
+
+// https://html.spec.whatwg.org/multipage/tables.html#table-processing-model
+
+export type Slot = { elements: Array<Element>, cell: Option<Cell> };
+function newSlot(cell: Cell | null = null): Slot {
+  return { elements: [], cell: Option.from(cell) }
+}
+
+export type Table = { slots: Array<Array<Slot>>, width: number, height: number, cells: Array<Cell>, rowGroups: Array<RowGroup>, colGroups: Array<ColGroup> };
+
+function getSlot(table: Table, col: number, row: number): Option<Slot> {
+  return Option
+    .from(table.slots[col])
+    .flatMap(sLine => Option.from(sLine[row]))
+}
+
+function setSlot(table: Table, col: number, row: number, slot: Slot) {
+  if (table.slots[col] === undefined) table.slots[col] = [];
+  table.slots[col][row] = slot;
+}
+
+export function newTable(): Table {
+  return { slots: [[]], width: 0, height: 0, cells: [], rowGroups: [], colGroups: []}
+}
+
+// Bad global variables! Bad!
+const global = { yCurrent:0};
```

Comment:
Either move to table class or have a specific "table under construction" class?

Reply:
I'd encapsulate all the state in an object that is passed around between the different functions.